File: leet_DP_583_two_strs_min_dist___LCS.py
class Solution:
    def minDistance(self, s1: str, s2: str) -> int:
        n1, n2 = len(s1), len(s2)
        dp = [[0] * (n2 + 1) for _ in range(n1 + 1)]
        for i in range(1, n1 + 1):
            for j in range(1, n2 + 1):
                if s1[i - 1] == s2[j - 1]:
                    dp[i][j] = 1 + dp[i - 1][j - 1]
                else:
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
        return n1 + n2 - 2 * dp[n1][n2]

File: leet_DP_583_two_strs_min_dist___LCS.cpp
class Solution {
public:
    int minDistance(string s1, string s2)
    {
        int                 n1 = s1.length(), n2 = s2.length(), i = 0, j;
        vector<vector<int>> dp(n1 + 1, vector<int>(n2 + 1, 0));

        while (++i < n1 + 1)
        {
            j = 0;
            while (++j < n2 + 1)
            {
                if (s1[i - 1] != s2[j - 1])
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
                else
                    dp[i][j] = 1 + dp[i - 1][j - 1];
            }
        }
        return (n1 + n2 - 2 * dp[n1][n2]);
    }
};
